query-engine/core/src/telemetry/capturing/capturer.rs

```diff
@@ -156,70 +112,199 @@ impl SpanProcessor for Processor {
+    use tokio::{sync::oneshot, time::interval};
+
+    const VALID_QUERY_ATTRS: [&str; 4] = ["query", "params", "target", "duration_ms"];
+    /// A Candidate represents either a span or an event that is being considered for capturing.
+    /// A Candidate can be converted into a [`Capture`].
+    #[derive(Debug, Clone)]
+    struct Candidate<'batch_iter> {
+        value: models::LogEvent,
+        settings: &'batch_iter Settings,
+    }
+
+    impl Candidate<'_> {
+        #[inline(always)]
+        fn is_loggable_query_event(&self) -> bool {
+            if self.settings.included_log_levels.contains("query") {
+                if let Some(target) = self.value.attributes.get("target") {
+                    if let Some(val) = target.as_str() {
+                        return (val == "quaint::connector::metrics" && self.value.attributes.get("query").is_some())
+                            || val == "mongodb_query_connector::query";
+                    }
                 }
             }
+            false
         }
-        false
-    }
 
-    fn query_event(mut self) -> models::LogEvent {
-        self.value
-            .attributes
-            .retain(|key, _| (&VALID_QUERY_ATTRS).contains(&key.as_str()));
+        fn query_event(mut self) -> models::LogEvent {
+            self.value
+                .attributes
+                .retain(|key, _| (&VALID_QUERY_ATTRS).contains(&key.as_str()));
 
-        models::LogEvent {
-            level: "query".to_string(),
-            ..self.value
+            models::LogEvent {
+                level: "query".to_string(),
+                ..self.value
+            }
         }
-    }
 
-    #[inline(always)]
-    fn is_loggable_event(&self) -> bool {
-        self.settings.included_log_levels.contains(&self.value.level)
+        #[inline(always)]
+        fn is_loggable_event(&self) -> bool {
+            self.settings.included_log_levels.contains(&self.value.level)
+        }
     }
-}
-
-/// tests for capture exporter
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use std::time::Duration;
 
-    #[tokio::test]
-    async fn test_garbage_collection() {
-        let exporter = Processor::new();
+    #[derive(Debug)]
+    pub(super) enum CaptureOp {
+        /// Tells the task that the given span data has been processed by the span processor
+        SpanDataProcessed(SpanData),
+        /// Tells the task to start capturing for the given trace id
+        StartCapturing(TraceId, Settings, oneshot::Sender<()>),
+        /// Tells the task to fetch the captures for the given trace_id, and sendthem to the given sender
+        FetchCaptures(TraceId, oneshot::Sender<Storage>),
+        /// Tells the task to cleaup traces that could have potentially expired
+        GC(TraceId),
+    }
 
-        let trace_id = TraceId::from_hex("1").unwrap();
-        let one_ms = Duration::from_millis(1);
+    static TASK: Lazy<Sender<CaptureOp>> = Lazy::new(|| {
+        let (sender, receiver) = unbounded();
```

Comment:
This is something I put in the prototype, but I am wondering if we are fine with an unbounded channel or if we want some backpressure. Not a blocker, putting the code in production is probably the best way to know.

Reply:
I don't think we need a backpressure ATM, no. we will see memory increase if this is problematic and also CPU cache invalidation in case the underlying data structure modeling the buffer queue needs reallocation. If we introduced bounded channels, we would need to measure the optimal channel capacity. This might be something that we will need, but better to do it when we have some load to check against.

I will follow up with @mavilein on the areas to observe to characterize the performance impact of query capturing.